Use jsonschema to constrain required or specific tool choice

TJ5 opened the pull request:
<!-- Thank you for your contribution! Please follow these guidelines to enhance your pull request. If anything is unclear, submit your PR and reach out to maintainers for assistance. Join our Slack community at https://slack.sglang.ai to discuss further. -->

## Motivation

EBNF constraints for tool choice are hard to read and extend. They must be individually created for each model's tool call format. Instead, we can use a much more easily extended json schema constraint. Json schema enables a much more simple constraint for required or named tool choices through passing constraints such as `minItems: 1`, which very easily enforces at least one tool call. The model output is constrained to json, so it can also bypass a parser for non-stream requests entirely. 

## Modifications

 - Created `get_json_schema_constraint`, a util function which outputs a json schema given the incoming tools and tool choice.
 - Created `json_array_parser`, which parses streaming output from requests for which tool_choice is set to 'required' or a named tool. 
 - Modified non-stream tool requests in `_process_tool_calls` to skip using a parser entirely and just extract the call information directly from the json output.
 - Added validations: Tools should specify a name and parameters. Parameters should be a valid json schema. If a tool choice is named to point to a specific tool, make sure that tool actually exists. We return a **400 BadRequestError** if any of these aren't true.
 - Updated the `_partial_json_loads()` function used by the base streaming parser. Since the incoming chunks for tool calls are an array, we have to also catch IndexErrors as well as JsonDecodeErrors to correctly load them and avoid a crash.
 - Note: for all json schemas, we use an **array schema**. If `tool_choice` is `required`, we specify `minItems: 1` so that at least one tool call is returned. If the tool_choice is a specific function, we specify `minItems: 1` and `maxItems: 1` so that exactly one tool call is returned, and it is the desired function per the tool choice.
 - Fix OpenAI protocl: `Function.name` should be required.


## Note on whitespace

Certain models (mistral, llama4) seem to generate excess whitespace in their tool calls with json schema. See: https://github.com/sgl-project/sglang/issues/8250 . Mistral in particular fails one of the existing test_tool_choice tests due to creating enough newline characters to hit the maximum token limit. Patching https://github.com/sgl-project/sglang/pull/8919 fixes this.

## Tests

### Tool Choice tests
   - Updated `test_error_handling_invalid_tool_choice test` to expect a 400 BadReqestError if we try to choose an invalid function
   - Added `test_invalid_tool_missing_name`, `test_invalid_tool_missing_parameters`, and `test_invalid_json_schema_in_tool` to expect a 400 BadRequestError if the user-submitted tools are malformed. For example: 
   
```python
{
                "type": "function",
                "function": {
                    "name": "test_function",
                    "description": "Test function with invalid JSON schema",
                    "parameters": {
                        "type": "object",
                        "properties": {
                            "invalid_field": {
                                "type": "unknown_type",  # Invalid type
                                "description": "This field has an invalid type",
                            }
                        },
                        "required": ["invalid_field"],
                    },
                },
            } 
```
  This isn't a valid json schema and will return a 400
   - The existing tests, which test a correct response for required tool choice requests, named tool choice requests, streaming and non-streaming also cover the json-schema constrained generation.
 - **Json schema tests**
   - New test file test_json_schema_constraints
   - Generates a variety of json schema constraints and verifies that they are both valid and accurately constrain the model. 
   - Also compares with the old ebnf constraint
 - **Function Call Parser Tests**
   - Added a class of tests to the existing function call parser tests to test the json array parser, which is used for streaming required/named tool choice requests
   - Test with malformed json, empty inputs, whitespace handling, separators in the middle of and at the end of chunks, and with multiple tool calls
   - Examples:
      - whitespace: chunk1: `' \n\n [{"name": "get_weather", "parameters": '`, chunk2: `'{"location": "Tokyo"}}]'`
      - malformed json: `'[{"name": "get_weather", "parameters": {"location": "Tokyo"'`

### Other tests done

Tested on various models:
| Model             | Required | Named Tool | Required Multiple Calls | Named Streaming | Required Streaming | Required Multiple Streaming |
|-------------------|----------|------------|-------------------------|-----------------|--------------------|-----------------------------|
| Deepseek V3       | ✅        | ✅          | ✅                       | ✅               | ✅                  | ✅                           |
| Deepseek V3.1     | ✅        | ✅          | ✅                       | ✅               | ✅                  | ✅                           |
| Qwen3             | ✅        | ✅          | ✅                       | ✅               | ✅                  | ✅                           |
| Qwen3-Coder       | ✅        | ✅          | ✅                       | ✅               | ✅                  | ✅                           |
| Llama-3.1-8B      | ✅        | ✅          | ✅                       | ✅               | ✅                  | ✅                           |
| Llama-4-Scout     | ✅        | ✅          | ✅                       | ✅               | ✅                  | ✅                           |
| Llama-4-Maverick  | ✅        | ✅          | ✅                       | ✅               | ✅                  | ✅                           |
| Mistral-v0.3      | ✅        | ✅          | ⚠️ (whitespace)          | ✅               | ✅                  | ⚠️ (whitespace)              |
| GPT-OSS-120B      | ✅        | ✅          | ✅                       | ✅               | ✅                  | ✅                           |


#### Deepseek-v3 Know Issue
 - Attempted to reproduce the tool calling bug reported for deepseekv3 here: https://github.com/sgl-project/sglang/issues/9311
 - Used the reporter's script. 
 
Output: (top: this PR, bottom: **main brach**) 
 - **Behavior:** Generated output contains two tool calls
 - The correct response has three tool calls, not two. However, the initial bug reported where the tool calls were malformed with empty URL fields is not present. 
 - The current main branch only returned one tool call.

<img width="1005" height="880" alt="Screenshot 2025-09-25 at 4 59 19 PM" src="https://github.com/user-attachments/assets/aa5e2ea7-4184-48d5-8105-7d4ab3d426ed" />
<img width="1012" height="708" alt="Screenshot 2025-09-25 at 4 59 40 PM" src="https://github.com/user-attachments/assets/194abd39-ae48-4584-b8fa-f5aff53f34dd" />

#### gpt-oss-120b Test
 - Also tested on gpt-oss-120b using the same script above. We expect 3 tool calls, get only one though. This is not an issue to do with this PR however as the same output is given if tool call is optional, using the old gpt-oss detector.

Output (for both tool_choice: required and auto): 
<img width="997" height="702" alt="Screenshot 2025-09-25 at 5 01 27 PM" src="https://github.com/user-attachments/assets/ed37525c-bec6-457b-bc38-b9dd6e008b2a" />



## Benchmarking and Profiling

<!-- If this pull request impacts inference speed, provide benchmarking and profiling results. -->

## Checklist

- [X] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [X] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

TJ5 committed:
squash

TJ5 committed:
more_tests

TJ5 committed:
ebnf comparison

TJ5 committed:
pre-commit

TJ5 committed:
validate tools

TJ5 commented on python/sglang/srt/entrypoints/openai/serving_chat.py:
```diff
@@ -952,13 +1030,31 @@ async def _process_tool_call_stream(
+        # Handle both FunctionCallParser and JsonArrayParser
+        if isinstance(parser, JsonArrayParser):
+            result = parser.parse_streaming_increment(delta, request.tools)
+            normal_text, calls = result.normal_text, result.calls
```
It does but in this case parse_streaming_increment is returning a `StreamingParseResult` object, not a tuple

CatherineSue commented on python/sglang/srt/function_call/json_array_parser.py:
```diff
@@ -0,0 +1,139 @@
+        # when we see the next JSON object start
+        if self._pending_separator:
+            # Check if the text (after removing leading whitespace) starts with '{'
+            if new_text.lstrip().startswith("{"):
```
What if I have a json string like `{"key": "value with } brace"}`?

CatherineSue commented on test/srt/function_call/test_json_schema_constraint.py:
```diff
@@ -0,0 +1,318 @@
+"""
```
Please add this to run_suite.py

TJ5 committed:
remove unneeded json.loads

TJ5 committed:
rework parser and more tests

TJ5 committed:
simplify logic

CatherineSue committed:
Remove MalformedJSON try-catch block

MalformedJSON is actually a ValueError in partial_json_parser. However, `_partial_json_loads` in `utils.py` only raise JSONDecodeError. So the try-catch there is never executed. Remove it.

CatherineSue committed:
Format files

CatherineSue committed:
Fix `_partial_json_loads` to handle whitespace

`JSONDecoder` can not trim JSON whitespace when doing `raw_decode`, add the handling to `_partial_json_loads`

CatherineSue committed:
Revert "Remove MalformedJSON try-catch block"

This reverts commit 8d88eea753d0527a8709dc890023d6b8153b4f0d.

CatherineSue committed:
Add back MalformedJSON import

CatherineSue committed:
Use WHITESPACE from json.decoder instead and format files

TJ5 committed:
fix partial_json_loads crash on ] character

TJ5 committed:
feedback

TJ5 commented on python/sglang/srt/function_call/json_array_parser.py:
```diff
@@ -0,0 +1,123 @@
+            "EBNF generation is not supported for JSON schema constraints."
+        )
+
+    def parse_streaming_increment(
```
Rewrote this whole section to just use the base class's parser without changes. Modified _partial_json_loads() to get around issues with whitespace and with the ']' character terminating certain chunks. All of these cases have been added to test_function_call_parser as well, and all pass.

TJ5 commented on python/sglang/srt/function_call/json_array_parser.py:
```diff
@@ -0,0 +1,139 @@
+                full_json = full_json[: -len(self.eot_token)]
+
+            try:
+                json.loads(full_json)
```
Repeat of other comment, but have overhauled this whole section so we are just using the base parser now. Modified _partial_json_loads() to get around previous issues with whitespace and with chunks terminating with the ']' character.

TJ5 committed:
clean up

TJ5 committed:
feedback and validate json schema

TJ5 committed:
fix test

TJ5 committed:
update tests

CatherineSue commented on python/sglang/srt/entrypoints/openai/serving_chat.py:
```diff
@@ -188,6 +217,19 @@ def _process_messages(
+                request.tool_choice == "required"
+                or isinstance(request.tool_choice, ToolChoice)
+                or (
+                    isinstance(request.tool_choice, dict)
```
Is dict needed here?

CatherineSue commented on python/sglang/srt/entrypoints/openai/serving_chat.py:
```diff
@@ -952,13 +1034,31 @@ async def _process_tool_call_stream(
+                    and "function" in request.tool_choice
+                )
+            ):
+                # Store JsonArrayParser directly for JSON schema constraints
```
please remove this comment as it is obivious

CatherineSue commented on python/sglang/srt/function_call/utils.py:
```diff
@@ -50,3 +55,83 @@ def _is_complete_json(input_str: str) -> bool:
+                defs = params.pop("$defs", {})
+                for def_name, def_schema in defs.items():
+                    if def_name in all_defs and all_defs[def_name] != def_schema:
+                        raise ValueError(
```
Who is responsible for catching this ValueError? Based on the code flow, this ValueError will be captured here: https://github.com/sgl-project/sglang/blob/c3faf2d6e673650b275bbefad9070706efddf3d3/python/sglang/srt/entrypoints/openai/serving_base.py#L65-L71

This is a 500 instead of 400. Is that intended?

TJ5 replied:
It's not. What I did to change this was I moved the validation to its own util function and call that with the rest of the validation logic in serving_chat.py.

TJ5 commented on python/sglang/srt/entrypoints/openai/serving_chat.py:
```diff
@@ -74,6 +78,27 @@ def _validate_request(self, request: ChatCompletionRequest) -> Optional[str]:
+                    return f"Tool {i} function is missing required 'name' field."
+                if (
+                    not hasattr(tool.function, "parameters")
+                    or tool.function.parameters is None
```
I'll fix the error message. Answering your earlier question, I saw it should be a json schema from openai's documentation on function calling:
<img width="1432" height="734" alt="image" src="https://github.com/user-attachments/assets/fb1941a5-263c-46bc-ae43-811f21ca3586" />
https://platform.openai.com/docs/guides/function-calling

TJ5 replied:
We don't need to check if we need a "function" field (I haven't pushed this change yet to get rid of the check, as of writing this) but pydantic doesn't catch if the parameters field doesn't exist. With that said, the Draft Schema validator also covers this case, so we don't need to check with `hasattr(tool.functino, "parameters")`

TJ5 committed:
feedback 1

TJ5 committed:
feedback 2

Swipe4057 commented:
Hi! Thanks for this awesome merge request! If it's not too much trouble, could you point me to which function or code is most likely to have bugs in gpt-oss function calling? (I'm new to the sglang codebase)

I wrote about the issue here:
https://github.com/sgl-project/sglang/issues/10738

I'm trying to fix it, but haven't been able to isolate the problem yet, and the current gpt-oss tests completely miss function calling bugs.

CatherineSue commented on python/sglang/srt/entrypoints/openai/serving_chat.py:
```diff
@@ -74,6 +82,34 @@ def _validate_request(self, request: ChatCompletionRequest) -> Optional[str]:
+                request.tool_choice, ToolChoice
+            ):
+                try:
+                    validate_tool_definitions(request.tools)
```
qq: this function is called for both required and ToolChoice here. But it is only called for required in utils.py. Is it intended?

CatherineSue commented on python/sglang/srt/function_call/utils.py:
```diff
@@ -50,3 +55,98 @@ def _is_complete_json(input_str: str) -> bool:
+        defs = params.pop("$defs", {})
+        for def_name, def_schema in defs.items():
+            if def_name in all_defs and all_defs[def_name] != def_schema:
+                raise ValueError(
```
Could we explore if there is a way to make this more concise? Right now this function is almost the same as `get_tool_schema_defs`. Just without ValueError raise.

CatherineSue commented:
Regarding the test_multi_tool_scenario_required for mistral, maybe mark it as optional or flaky for now.

TJ5 committed:
condense defs logic and add test cases

TJ5 committed:
pre-commit

CatherineSue commented on python/sglang/srt/entrypoints/openai/serving_chat.py:
```diff
@@ -867,14 +918,59 @@ def _process_tool_calls(
+        # Use parser since output is not constrained by JSON schema
         parser = FunctionCallParser(tools, self.tool_call_parser)
         if parser.has_tool_call(text):
+            # Set finish reason to tool_calls since we detected tool calls
```
let's remove this.

CatherineSue commented on python/sglang/srt/entrypoints/openai/serving_chat.py:
```diff
@@ -867,14 +918,59 @@ def _process_tool_calls(
             if finish_reason["type"] == "stop":
                 finish_reason["type"] = "tool_calls"
                 finish_reason["matched"] = None
+
```
remove this line as well. Keep minimal changes

TJ5 committed:
validate with pydantic and skip broken mistral test

TJ5 committed:
feedback 3

TJ5 committed:
skip mistral whitespace-failing test

CatherineSue commented on python/sglang/srt/entrypoints/openai/serving_base.py:
```diff
@@ -62,6 +62,21 @@ async def handle_request(
                 message=e.detail, err_type=str(e.status_code), status_code=e.status_code
             )
+        except ValueError as e:
+            # Check if this is a tool schema conflict error
```
ValueError can be treated as 400 in general.

Let's remove the check and just do:

```
                return self.create_error_response(
                    message=str(e),
                    err_type="BadRequest",
                    status_code=400,
                )
```

CatherineSue commented:
Overall LGTM. Let's wait for the CI results. Then I will approve.

TJ5 committed:
return 400 on all valueerrors

CatherineSue commented on python/sglang/srt/entrypoints/openai/serving_chat.py:
```diff
@@ -75,6 +80,23 @@ def _validate_request(self, request: ChatCompletionRequest) -> Optional[str]:
+            if not tool_exists:
+                return f"Tool '{tool_name}' not found in tools list."
+
+        # Validate tool definitions
```
This might reduce the indentation more.

```suggestion
for i, tool in enumerate(request.tools or []):
    if tool.function.parameters is None:
        continue
    try:
        Draft202012Validator.check_schema(tool.function.parameters)
    except SchemaError as e:
        return f"Tool {i} function has invalid 'parameters' schema: {e}"
```

CatherineSue replied:
Note: we can't commit directly as I didn't wrap all the lines.

TJ5 committed:
reduce indents

CatherineSue commented:
unit-test-backend-1-gpu(5) is failing due to HF 403: https://github.com/sgl-project/sglang/actions/runs/18052643506/job/51379091807?pr=10550

JustinTong0323 commented:
Note: The only not passed CI is failed due to an error on main, fix in #10977
